### PR TITLE
v5.0.0 - Notifications

### PR DESCRIPTION
**Pull request by cdechaux**
Adding notifications for all in HomePage and a new page for managing notifications for Admin and Assos.

Back :

- New model file notification.py in betterave_backend/app/models
- New operation file notification_operations.py in betterave_backend/app/operations
- API : new folder notifications in betterave_backend/app/api with 4 new files (init.py, namespace.py, models.py, routes.py)
- Adding routes for notifications in betterave_backend/app/api/users.route.py
- New relationships in betterave_backend/app/models/relationship_tables.py and in betterave_backend/app/models/user.py
- Modifications to add notifications in betterave_backend/app/operations/user_operations.py
- Adding notifications at the beginning in betterave_backend/app/scripts/init_db.py
- New test file betterave_backend/tests/test_notification_operations.py

Front :

- New component Notification.vue in frontend/src/components for the Home Page
- New controls NotificationControls.vue frontend/src/views/controls -> to manage notifications
- New component CreateNotification.vue in frontend/src/components -> to add and remove notifications for Admins and Assos
- Modifications on ProfilPill.vue in frontend/src/components to add the previous page to menu for Admins and Assos
- Modification of frontend/src/views/HomePage.vue to add visualization of notifications in Home page

att F9MJY3bFk0RlWBElO6p60iYhFL1w3WG-sw2P78B3cQY
betterave_2

### CODE REVIEW

File: betterave-backend/betterave_backend/app/operations/notification_operations.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+from betterave_backend.extensions import db
+from betterave_backend.app.models import Notification, User, UserLevel
+from betterave_backend.app.decorators import is_valid_apikey, with_instance
+
+
+def add_notification(
+    title: str,
+    content: str,
+    sent_by_user_id: int,
+    recipient_type,
```
Comment (thomktz) [inline betterave-backend/betterave_backend/app/operations/notification_operations.py:14]: type hint: Union[str, UserLevel]

File: betterave-backend/betterave_backend/app/operations/notification_operations.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+) -> int:
+    """Add a notification to the database."""
+    try:
+        if recipient_type == "Subscribers":
+            recipient_users = User.query.get(sent_by_user_id).subscribers
+        elif recipient_type == "All users":
+            recipient_users = User.query.all()
+        else:
```
Comment (thomktz) [inline betterave-backend/betterave_backend/app/operations/notification_operations.py:22]: Maybe use `elif isinstance(recipient_type, UserLevel)` and throw error on `else`

File: betterave-backend/betterave_backend/app/operations/notification_operations.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        db.session.commit()
+
+        return new_notification.notification_id
+    except SQLAlchemyError as e:
+        db.session.rollback()
+        print(f"Error adding notification: {str(e)}")
+        return -1
+
+
+def update_notification(notification_id, new_data: dict) -> bool:
```
Comment (thomktz) [inline betterave-backend/betterave_backend/app/operations/notification_operations.py:44]: type hint `notification_id`

File: betterave-backend/betterave_backend/app/operations/notification_operations.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+
+def get_notification_by_id(notification_id: int) -> Notification:
+    """Get a notification by its ID."""
+    return db.session.get(Notification, notification_id)
+
+
+def get_all_notifications() -> list[Notification]:
+    """Return all notifications in the database."""
+    # return [db.session.get(Notification,3)]
```
Comment (thomktz) [inline betterave-backend/betterave_backend/app/operations/notification_operations.py:83]: Remove comment

File: betterave-backend/betterave_backend/app/operations/notification_operations.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+    return notif.title
+
+
+@with_instance(User)
+def get_user_notifications(user: User, limit: Optional[int] = None) -> list[Notification]:
+    """Get all events a particular user is attending."""
+    if limit is not None:
+        notification = user.receptionned_notifications[:limit]
+    else:
+        notification = user.receptionned_notifications
```
Comment (thomktz) [inline betterave-backend/betterave_backend/app/operations/notification_operations.py:99]: `User.received_notifications` is probably a better name for the attribute

File: betterave-backend/betterave_backend/app/operations/notification_operations.py [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
+    return notification
+
+
+def add_recipient_to_notification(notification_id, user_ids=None, user_level=None):
```
Comment (thomktz) [inline betterave-backend/betterave_backend/app/operations/notification_operations.py:104]: Type hint

File: betterave-backend/betterave_backend/app/operations/user_operations.py [hunk 7]
```diff
@@ -153,6 +160,37 @@ def update_event_attendance(user: User) -> None:
...
+    for notif in get_all_notifications():
+        # If user already has an attendance for this event
+        if notif in user.attended_events:
+            if notif.recipient_type == "Subscribers" and notif.association not in user.subscriptions:
+                user.receptionned_notifications.remove(notif)
+
+            elif notif.recipient_type == "All users":
+                continue
+
+            if notif.recipient_type != user.level.value:
```
Comment (thomktz) [inline betterave-backend/betterave_backend/app/operations/user_operations.py:176]: Same, `elif isinstance(...): ..., else: remove anyways, or skip`

File: betterave-backend/tests/test_event_operations.py [hunk 8]
```diff
@@ -27,6 +27,8 @@
 LOCATION = "Amphi 200"
 PARTICIPANT_TYPE = "All users"
 
+print("test event test")
```
Comment (thomktz) [inline betterave-backend/tests/test_event_operations.py:30]: To remove

**Review by thomktz — COMMENTED** [review]
Très propre! Est-ce que tu peux ajouter un screenshot dans la description du PR, et ajouter une entrée dans le CHANGELOG.md sous une version `## - [unreleased]` ?

**Comment by emanieluu** [comment]
Super clair 👍

**Review by thomktz — APPROVED** [review]

**Pull request merged**